tools/src/main/java/org/apache/kafka/tools/ConnectPluginPath.java

```diff
@@ -0,0 +1,518 @@
+                return new Config(Command.LIST, locations, out);
+            default:
+                throw new ArgumentParserException("Unrecognized subcommand: '" + subcommand + "'", parser);
+        }
+    }
+
+    private static Set<Path> parseLocations(ArgumentParser parser, Namespace namespace) throws ArgumentParserException, TerseException {
+        List<String> rawLocations = new ArrayList<>(namespace.getList("plugin_location"));
+        List<String> rawPluginPaths = new ArrayList<>(namespace.getList("plugin_path"));
+        List<String> rawWorkerConfigs = new ArrayList<>(namespace.getList("worker_config"));
+        if (rawLocations.isEmpty() && rawPluginPaths.isEmpty() && rawWorkerConfigs.isEmpty()) {
+            throw new ArgumentParserException("Must specify at least one --plugin-location, --plugin-path, or --worker-config", parser);
+        }
+        Set<Path> pluginLocations = new LinkedHashSet<>();
+        for (String rawWorkerConfig : rawWorkerConfigs) {
+            Properties properties;
+            try {
+                properties = Utils.loadProps(rawWorkerConfig);
+            } catch (IOException e) {
+                throw new TerseException("Unable to read worker config at " + rawWorkerConfig);
+            }
+            String pluginPath = properties.getProperty(WorkerConfig.PLUGIN_PATH_CONFIG);
+            if (pluginPath != null) {
+                rawPluginPaths.add(pluginPath);
+            }
+        }
+        for (String rawPluginPath : rawPluginPaths) {
+            try {
+                pluginLocations.addAll(PluginUtils.pluginLocations(rawPluginPath, true));
+            } catch (UncheckedIOException e) {
+                throw new TerseException("Unable to parse plugin path " + rawPluginPath + ": " + e.getMessage());
+            }
+        }
+        for (String rawLocation : rawLocations) {
+            Path pluginLocation = Paths.get(rawLocation);
+            if (!pluginLocation.toFile().exists()) {
+                throw new TerseException("Specified location " + pluginLocation + " does not exist");
+            }
+            pluginLocations.add(pluginLocation);
+        }
+        return pluginLocations;
+    }
+
+    private static Path makeDistinct(Path path) {
+        // When the same jar is present on the classpath and inside a pluginLocation, we still need to be able
+        // to distinguish between them. In order to make them more likely to be distinct, change the path to
+        // include an arbitrary change that makes them still resolve the same file, but fail direct equality checks.
+        Path parent = path.getParent();
+        if (parent == null) {
+            // At the root of the filesystem, add the current directory marker to the end.
+            return path.resolve(".");
+        } else {
+            // If we have a parent directory, add the current directory marker there.
+            return parent.resolve(".").resolve(path.getFileName());
+        }
+    }
```

Comment:
This feels like a bit of a hack, and may make debugging and/or future modifications more difficult.

What do you think about adding a `boolean isolated` field to `ManifestEntry` that can be used to distinguish between classpath and non-classpath plugins?

Reply:
> What do you think about adding a boolean isolated field to ManifestEntry that can be used to distinguish between classpath and non-classpath plugins?

Unfortunately this doesn't help because the ambiguity happens before we construct the ManifestEntry. Here's the situation:

* `/path/to/X.jar` contains manifest file named Y.
* The classpath and at least one pluginLocation lists Jar X.
* `AppClassLoader#getResources(Y)` returns `[jar:file:/path/to/X.jar!/Y]` which then becomes one ManifestEntry.
* `PluginClassLoader#getResources(Y)` returns `[jar:file:/path/to/X.jar!/Y, jar:file:/path/to/X.jar!/Y]` which then becomes one ManifestEntry due to Set equality.

My previous solution was to change `PluginClassLoader#getResources(Y)` to return `[jar:file:/path/to/X.jar!/Y, jar:file:/path/to/./X.jar!/Y]` so that it would become two different ManifestEntry objects.

The new solution keeps the original `PluginClassLoader#getResources(Y)` which returns `[jar:file:/path/to/X.jar!/Y, jar:file:/path/to/X.jar!/Y]`, but creates two identical ManifestEntry objects in a List. Then when processing the excludes, one copy is removed and the other stays behind.